src/main/java/org/jenkinsci/plugins/junitrealtimetestreporter/RealtimeJUnitStep.java

```diff
@@ -237,7 +242,6 @@ private void finished(StepContext context, boolean success) throws Exception {
                 pipelineTestDetails.setEnclosingBlocks(JUnitResultsStepExecution.getEnclosingBlockIds(enclosingBlocks));
                 pipelineTestDetails.setEnclosingBlockNames(JUnitResultsStepExecution.getEnclosingBlockNames(enclosingBlocks));
 
-                // TODO might block CPS VM thread. Not trivial to solve: JENKINS-43276
```

Comment:
No, this is still blocking the CPS VM thread. You would need to wrap `finished` in `run` I think.

Reply:
Hmm I've had a go at it, something isn't working right when it's async.

I had one run in the UI complete before it could set the job unstable, generally it seems to work though.

But the tests fail with incorrect test counts.

![image](https://user-images.githubusercontent.com/21194782/148032002-3e10d3fe-8627-4ff7-b509-9b962c529397.png)

Patch applied to test:

```gitdiff
diff --git a/src/test/java/org/jenkinsci/plugins/junitrealtimetestreporter/RealtimeJUnitStepTest.java b/src/test/java/org/jenkinsci/plugins/junitrealtimetestreporter/RealtimeJUnitStepTest.java
index 5ca2cce..e84fc81 100644
--- a/src/test/java/org/jenkinsci/plugins/junitrealtimetestreporter/RealtimeJUnitStepTest.java
+++ b/src/test/java/org/jenkinsci/plugins/junitrealtimetestreporter/RealtimeJUnitStepTest.java
@@ -195,7 +195,10 @@ public class RealtimeJUnitStepTest {
                     "    }\n" +
                     "  }\n" +
                     "}", true));
-                WorkflowRun b1 = rr.j.assertBuildStatus(Result.UNSTABLE, p.scheduleBuild2(0).get());
+                WorkflowRun run = p.scheduleBuild2(0).get();
+                rr.j.waitForCompletion(run);
+                Thread.sleep(10000);
+                WorkflowRun b1 = rr.j.assertBuildStatus(Result.UNSTABLE, run);
```

```
[ERROR] Tests run: 5, Failures: 1, Errors: 0, Skipped: 0, Time elapsed: 126.695 s <<< FAILURE! - in org.jenkinsci.plugins.junitrealtimetestreporter.RealtimeJUnitStepTest
[ERROR] org.jenkinsci.plugins.junitrealtimetestreporter.RealtimeJUnitStepTest.testResultDetails  Time elapsed: 16.832 s  <<< FAILURE!
java.lang.AssertionError: expected:<4> but was:<2>
	at org.junit.Assert.fail(Assert.java:89)
	at org.junit.Assert.failNotEquals(Assert.java:835)
	at org.junit.Assert.assertEquals(Assert.java:647)
	at org.junit.Assert.assertEquals(Assert.java:633)
	at org.jenkinsci.plugins.junitrealtimetestreporter.RealtimeJUnitStepTest.lambda$testResultDetails$4(RealtimeJUnitStepTest.java:204)
	at org.jvnet.hudson.test.RestartableJenkinsRule$3.evaluate(RestartableJenkinsRule.java:243)
	at org.jvnet.hudson.test.RestartableJenkinsRule$6.evaluate(RestartableJenkinsRule.java:291)
	at org.jvnet.hudson.test.JenkinsRule$1.evaluate(JenkinsRule.java:601)
	at org.junit.internal.runners.statements.FailOnTimeout$CallableStatement.call(FailOnTimeout.java:299)
	at org.junit.internal.runners.statements.FailOnTimeout$CallableStatement.call(FailOnTimeout.java:293)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
	at java.base/java.lang.Thread.run(Thread.java:829)
```

Reply:
Not sure. What is the patch you used?

Reply:
it's https://github.com/jenkinsci/junit-realtime-test-reporter-plugin/pull/26/commits/8b3f06e2d1793c7d2f471c5825870b577e202999